attestation-service/docs/amd-offline-certificate-cache.md

```diff
@@ -0,0 +1,247 @@
+servicing. The machine's VCEK URL can be fetched using the
+[snphost tool](https://github.com/virtee/snphost).
+
+On each AMD SNP host, with root/admin access run:
+
+```
+snphost show vcek-url
+```
+
+You should get a URL of the form:
+
+```
+https://kdsintf.amd.com/vcek/{version}/{machine}/{product_name}/{hwid}?{params}
+```
+
+See the [AMD VCEK documentation](https://www.amd.com/content/dam/amd/en/documents/epyc-technical-docs/specifications/57230.pdf)
+for more information about the VCEK URL format.
+
+> [!IMPORTANT]
+> Note that the VCEK URL is specific to the hardware AMD firmware of the
+> machine. If the firmware is updated, the VCEK URL will change.
+
+### 2. Run cache_preloader to generate cache contents
+
+On a system with access to the AMD KDS, run `cache_preloader` specifying
+the VCEK URL file and output directory.
+
+```
+# Navigate to the cache-preloader directory
+cd tools/cache-preloader
+
+# Preload the cache and create an archive
+cargo run -- -u vcek_urls.txt -c ./vcek_cache -a vcek_cache.tar.gz
+```
+
+### 3. Configure the attestation service
+
+Update the attestation config file to use a predefined cache:
+
+`kbs/config/as-config.json`:
+
+```json
+{
+    // ... other fields ...
+    "verifier_config": {
+        "snp_verifier": {
+            "cache": {
+                "type": "HttpCacheReqwest",
+                "offline_mode": true,
+                "cache_dir": "/path/to/cache"
+            }
+        }
+    }
+}
+```
+
+### 4. Install cache contents on air-gapped attestation deployment
+
+Copy output directory contents to your air-gapped attestation deployment
+maintaining the same directory structure.
+
+### Docker:
+
+- **Install on running attestation service**
+
+You may copy the cache to the container after it has started - ensure that
+the as-config.json has picked up your expected cache directory:
+```
+$ docker logs trustee-as-1 | grep cache_dir
+                            cache_dir: "/etc/test-vcek-cache",
+```
+
+Then copy your archive (making sure to add the ending `/` in both paths):
+```
+docker cp ./tools/cache-preloader/vcek_cache/ trustee-as-1:/etc/test-vcek-cache/
+```
+
+- **Build into image**
+
+You may also copy the cache as a part of building the image by updating the
+attestation service's Docker file:
+
+```
++++ b/attestation-service/docker/as-grpc/Dockerfile
+@@ -63,6 +63,7 @@ RUN if [ "${VERIFIER}" = "all-verifier" ] || [ "${VERIFIER}" = "az-snp-vtpm-veri
+     rm -rf /var/lib/apt/lists/{apt,dpkg,cache,log} /tmp/* /var/tmp/*; fi
+
+ COPY --from=builder /usr/local/cargo/bin/grpc-as /usr/local/bin/grpc-as
++COPY ./attestation-service/vcek_cache /etc/test-vcek-cache
+```
+
+In the above example, the `cache-preloader` tool's target directory was `../../attestation-service/vcek_cache`:
+```
+cargo run -- -u test_urls.txt -c ../../attestation-service/vcek-cache
+```
+
+And `as-config.json` sets `cache_dir` to `/etc/test-vcek-cache`.
```

Comment:
Not totally canonical to put a cache in `/etc` imo. Also I'm not sure that baking the cached certs into the image itself really makes sense, but I guess it's fine.

Reply:
Yeah I think that would be the most common way to set it up